From: Hannes Reinecke <hare@suse.de>
Subject: SATA cdrom oopses

SATA CDROM again. We alvays have to call scsi_finish_command as this
routine is called from eh_strategy_handler which does not do _any_
postprocessing on scsi command.

--- linux-2.6.12/drivers/scsi/libata-core.c.orig	2005-08-25 09:54:12.000000000 +0200
+++ linux-2.6.12/drivers/scsi/libata-core.c	2005-08-25 10:08:19.000000000 +0200
@@ -2929,18 +2929,17 @@ static void ata_qc_timeout(struct ata_qu
 	if (qc->dev->class == ATA_DEV_ATAPI && qc->scsicmd) {
 		struct scsi_cmnd *cmd = qc->scsicmd;
 
-		if (!(cmd->eh_eflags & SCSI_EH_CANCEL_CMD)) {
-
-			/* finish completing original command */
-			__ata_qc_complete(qc);
+		/* finish completing original command */
+		__ata_qc_complete(qc);
 
+		if (!SCSI_SENSE_VALID(cmd)) {
 			atapi_request_sense(ap, dev, cmd);
-
+		
 			cmd->result = (CHECK_CONDITION << 1) | (DID_OK << 16);
-			scsi_finish_command(cmd);
-
-			goto out;
 		}
+		scsi_finish_command(cmd);
+
+		goto out;
 	}
 
 	/* hack alert!  We cannot use the supplied completion
